x/oracle/keeper/owasm.go

```diff
@@ -152,3 +160,42 @@ func (k Keeper) ResolveRequest(ctx sdk.Context, reqID types.RequestID) {
 		k.ResolveSuccess(ctx, reqID, env.Retdata, output.GasUsed)
 	}
 }
+
+// CollectFee subtract fee from fee payer and send them to treasury
+func (k Keeper) CollectFee(ctx sdk.Context, payer sdk.AccAddress, feeLimit sdk.Coins, askCount uint64, rawRequests []types.RawRequest) (sdk.Coins, error) {
+
+	if err := feeLimit.Validate(); err != nil {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidCoins, err.Error())
+	}
+
+	collector := newFeeCollector(k.bankKeeper, feeLimit, payer)
+
+	for _, r := range rawRequests {
+
+		ds, err := k.GetDataSource(ctx, r.DataSourceID)
+		if err != nil {
+			return nil, err
+		}
+
+		if ds.Fee.Empty() {
+			continue
+		}
+
+		fee := sdk.NewCoins()
+		for _, c := range ds.Fee {
+			c.Amount = c.Amount.Mul(sdk.NewInt(int64(askCount)))
+			fee = fee.Add(c)
+		}
+
+		treasury, err := sdk.AccAddressFromBech32(ds.Treasury)
+		if err != nil {
+			return nil, err
+		}
+
+		if err := collector.Collect(ctx, fee, treasury); err != nil {
+			return nil, err
```

Comment:
If we have an error for some raw request, so we can't collect the fee?

Reply:
should state be revert if an error is returned?